### PR TITLE
New idiom: Out of the Box Dynamic Dispatch

### PR DESCRIPTION
**Pull request by llogiq**

### CODE REVIEW

File: idioms/ootb-dyn-dispatch.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+## Description
+
+We can dynamically dispatch over multiple values, however, to do so, we need
+to declare multiple variables to bind differently-typed objects. To extend the
+lifetime as necessary, we can use deferred conditional initialization, as seen
+below:
+
+## Example
+
+``rust
```
Comment (0x20F) [inline idioms/ootb-dyn-dispatch.md]: Codeblocks start with three ```

**Comment by twe4ked** [comment]
Great write up! I was a bit confused by the title at first, perhaps something along the lines of “Dynamic Dispatch without Box<T>”?

**Comment by llogiq** [comment]
Perhaps on-stack dynamic dispatch?

**Comment by twe4ked** [comment]
That sounds good to me ☺️

File: README.md [hunk 2]
```diff
@@ -24,6 +24,7 @@ language.
 * [Pass variables to closure](idioms/pass-var-to-closure.md)
 * [`mem::replace(_)` to avoid needless clones](idioms/mem-replace.md)
 * [Temporary mutability](idioms/temporary-mutability.md)
+* [Out of the Box Dynamic Dispatch](idioms/ootb-dyn-dispatch.md)
```
Comment (Nashenas88) [inline README.md]: Isn't this title now out of sync with the title below?

Comment (llogiq) [reply README.md]: Right, fixed.

**Comment by llogiq** [comment]
It's been a while since there was any movement on this. Is there something I should do or is this good to merge?

**Comment by llogiq** [comment]
@lambda-fairy @Manishearth anyone?

**Comment by lambda-fairy** [comment]
Hey @llogiq!

Sorry, I've had way too many things going on lately 😢 

It looks like the document is still called `ootb-dyn-dispatch.md`. Can you please rename that to something like `on-stack-dyn-dispatch.md`?

Otherwise it looks good to me 👍

**Comment by burdges** [comment]
I suppose LLVM always eliminates the duplicate stack usage here, so there is never a reason to do this with a union.

**Comment by pickfire** [comment]
@llogiq I think you should also add an example for `Option`.

File: idioms/ootb-dyn-dispatch.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# On-Stack Dynamic Dispatch
```
Comment (pickfire) [inline idioms/ootb-dyn-dispatch.md:1]: I think this is more well-known as out-of-box dynamic dispatch from blog posts, maybe we could mention it as well?

Comment (llogiq) [reply idioms/ootb-dyn-dispatch.md:1]: I don't think that will be necessary.

**Comment by llogiq** [comment]
@pickfire I've noted in the "see also" section that `Option.as_ref()` solves most of the same problem for us.

**Comment by llogiq** [comment]
Moved the file and rebased. @lambda-fairy r?

**Pull request merged**